front/pages/[user]/a/[sId]/execute/index.jsx

```diff
@@ -0,0 +1,403 @@
+            <ChevronRightIcon className="h-4 w-4 mt-0.5" />
+          ) : (
+            <ChevronDownIcon className="h-4 w-4 mt-0.5" />
+          )}{" "}
+          {blockName}{" "}
+          {lastEventForBlock.content.status === "running" ? (
+            <Spinner />
+          ) : (
+            <CheckCircleIcon className="text-emerald-300 h-4 w-4 min-w-4 mt-0.5" />
+          )}
+        </div>
+      </Button>
+      {expanded ? (
+        <div className="flex ml-4">
+          {Array.isArray(preprocessedOutput) ? (
+            <ArrayViewer value={preprocessedOutput} />
+          ) : typeof preprocessedOutput == "string" ? (
+            <StringViewer value={preprocessedOutput} />
+          ) : typeof outputForBlock.content.execution == "object" ? (
+            <ObjectViewer value={preprocessedOutput} />
+          ) : null}
+        </div>
+      ) : null}
+    </>
+  );
+}
+
+function ExecuteOutput({ executionLogs, expandedByBlockName, onToggleExpand }) {
+  return executionLogs.blockOrder.length ? (
+    <>
+      {executionLogs.blockOrder.map((blockName, index) => {
+        const lastEventForBlock = executionLogs.lastEventByBlockName[blockName];
+        const outputForBlock = executionLogs.outputByBlockName[blockName];
+        return (
+          <ExecuteOutputLine
+            key={blockName}
+            blockName={blockName}
+            outputForBlock={outputForBlock}
+            lastEventForBlock={lastEventForBlock}
+            expanded={expandedByBlockName[blockName]}
+            onToggleExpand={() => onToggleExpand(blockName)}
+          />
+        );
+      })}
+      <br /> {/*TODO: no br*/}
+    </>
+  ) : null;
+}
+
+function ExecuteInput({ inputName, inputValue, onChange }) {
+  return (
+    <div key={inputName} className="grid grid-cols-10">
+      <div className="flex group items-center bg-slate-300">
+        <input
+          className={classNames(
+            "px-1 py-1 font-normal text-[13px] font-mono bg-slate-300 border-0 outline-none focus:outline-none w-full",
+            "border-white ring-0 focus:ring-0 focus:border-white"
+          )}
+          readOnly={true}
+          value={inputName}
+        />
+      </div>
+      <div
+        className={classNames(
+          "col-span-7 inline-grid space-y-0 resize-none text-[13px] font-mono px-0 py-0 border bg-slate-100",
+          "border-slate-100"
+        )}
+      >
+        <TextareaAutosize
+          minRows={1}
+          className={classNames(
+            "w-full resize-none font-normal text-[13px] font-mono px-1 py-0 bg-transparent border-0 ring-0 focus:ring-0",
+            "text-gray-700"
+          )}
+          value={inputValue || ""}
+          onChange={(e) => onChange(e.target.value)}
+        />
+      </div>
+    </div>
+  );
+}
+
+export default function ExecuteView({
+  app,
+  user,
+  ga_tracking_id,
+  inputDataset,
+}) {
+  const { data: session } = useSession();
+
+  // TODO: check idiomatic way to do this
```

Comment:
how do you typically deal with this ? I'd rather avoid redoing this work on every render

Reply:
Good question. Not sure this get rerendered that often since these is dependent only on server generated props?

Reply:
I think you are right, I'll remove the useStates

Reply:
Actually I ended up moving those server-side, which solves the problem.